functions/webhooks/serviceConversationListener.protected.ts

```diff
@@ -0,0 +1,125 @@
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+import { Context, ServerlessCallback } from '@twilio-labs/serverless-runtime-types/types';
+import { responseWithCors, bindResolve, error500 } from '@tech-matters/serverless-helpers';
+
+export type ConversationSid = `CH${string}`;
+export type ParticipantSid = `MB${string}`;
+
+type ServiceConversationListenerEvent = {
+  Body: string;
+  ParticipantSid: ParticipantSid;
+  ConversationSid: ConversationSid;
+  EventType: string;
+  MessageSid: string;
+};
+
+export type Body = ServiceConversationListenerEvent;
+
+export const sendConversationMessage = async (
+  context: Context,
+  {
+    conversationSid,
+    author,
+    messageText,
+    messageAttributes,
+  }: {
+    conversationSid: ConversationSid;
+    author: string;
+    messageText: string;
+    messageAttributes?: string;
+  },
+) =>
+  context
+    .getTwilioClient()
+    .conversations.conversations.get(conversationSid)
+    .messages.create({
+      body: messageText,
+      author,
+      xTwilioWebhookEnabled: 'true',
+      ...(messageAttributes && { attributes: messageAttributes }),
+    });
+
+const getTimeDifference = async (isoString: Date): Promise<string> => {
+  const unitCheck = (count: number, unit: string) => (count > 1 ? `${unit}s` : unit);
+  // Create a new Date object from the ISO string
+  const givenDate = new Date(isoString);
+
+  // Get the current date and time
+  const currentDate = new Date();
+
+  // Calculate the difference in milliseconds
+  const differenceInMillis = currentDate.getTime() - givenDate.getTime();
+
+  // Convert the difference to seconds
+  const differenceInSeconds = Math.floor(differenceInMillis / 1000);
+  const differenceInMinutes = Math.floor(differenceInSeconds / 60);
+
+  // Determine whether to return the difference in seconds or minutes
+  return differenceInSeconds < 60
+    ? `${differenceInSeconds} ${unitCheck(differenceInSeconds, 'second')}`
+    : `${differenceInMinutes} ${unitCheck(differenceInMinutes, 'minute')}`;
+};
+
+export const handler = async (context: Context, event: Body, callback: ServerlessCallback) => {
+  const response = responseWithCors();
+  const resolve = bindResolve(callback)(response);
+  try {
+    const { EventType, ConversationSid, MessageSid, ParticipantSid, Body } = event;
+
+    let messageAuthor: string | undefined;
+
+    // check if it's the onMessageAdded event
+    if (EventType === 'onMessageAdded') {
+      const conversationMessage = await context
```

Comment:
Can we move the logic for sending errors for unsupported media into a separate private function.  The top level webhook handler should just be a general conversation webhook handler that delegates to other functions for specific purposes. If there is a problem with the error messages in the future it will be much easier to find the relevant code if it is in a function module called something like `sendErrorMessageForUnsupportedMedia.private.ts` than if it just somewhere in the generic `serviceConversationListener.protected.ts`